jobrunner/project.py

```diff
@@ -234,6 +237,35 @@ def get_action_specification(project, action_id):
     if "config" in action_spec:
         run_command = add_config_to_run_command(run_command, action_spec["config"])
     run_args = shlex.split(run_command)
+
+    # Does run_args run a reusable action?
+    image, tag = run_args[0].split(":")
+    if image not in config.ALLOWED_IMAGES:
+        # Fetch the reusable action
+        repo_url = f"{config.ACTIONS_GITHUB_ORG}/{image}"
+        commit_sha = git.get_sha_from_remote_ref(repo_url, tag)
+        action_file = git.read_file_from_repo(repo_url, commit_sha, "action.yaml")
+        action = parse_and_validate_action_file(action_file)
+
+        # Modify run_args
+        # ["action:tag", "arg", ...]
+
+        run_args[0] = action["runtime"]
+        # ["runtime:tag", "arg", ...]
+
+        run_args.insert(1, action["entrypoint"])
+        # ["runtime:tag", "entrypoint", "arg", ...]
+
+        binary = get_binary_for_reusable_action(action)
+        if binary:
+            run_args.insert(1, binary)
+            # ["runtime:tag", "binary", "entrypoint", "arg", ...]
```

Comment:
They need not be; we could instead ask that *action.yaml* contains one property, possibly called `run`. For example:

```yaml
run: python:latest python reusable_action/main.py
```

We'd replace `run_args[0]` with `action["run"]` on line 253 above. Is that what you're suggesting? If so, then I'll pop a message in Slack, as I can't remember why we separated the elements. This approach would be better, though: it would make writing a reusable action even more like writing an action.

Reply:
Yes, effectively. Although I think to get the argument separation working it would need to be:
```py
run_args[:1] = shlex.split(action["run"])
```

Reply:
I'm going to push a commit that replaces `runtime`/`entrypoint` with `run`. We no longer need to assemble a replacement for `run_args[0]` and can remove `get_binary_for_reusable_action`. I don't think we need to separate the arguments, though. Possibly the word "action" is confusing, here?

```python
action = {"run": "reusable-action:latest --output-format=png"}  # From project.yaml
run_args = shlex.split(action["run"])
run_args  # ['reusable-action:latest', '--output-format=png']

action_config = {"run": "python:latest python reusable_action/main.py"}  # From action.yaml

run_args[0] = action_config["run"]  # No argument separation!
run_args  # ['python:latest python reusable_action/main.py', '--output-format=png']

action["run"] = " ".join(run_args)
action  # {'run': 'python:latest python reusable_action/main.py --output-format=png'}
```